File: apps/windows/src-windows/VisualTeX.WindowsOffice.Contracts/FormulaEquationTag.cs
using System;

namespace VisualTeX.WindowsOffice.Contracts;

public readonly struct FormulaEquationTagSplit
{
    public FormulaEquationTagSplit(string latex, string? equationTag)
    {
        Latex = latex;
        EquationTag = equationTag;
    }

    public string Latex { get; }
    public string? EquationTag { get; }
}

public static class FormulaEquationTag
{
    public static FormulaEquationTagSplit Extract(string? source)
    {
        var trimmed = (source ?? string.Empty).Trim();
        if (trimmed.Length == 0) return new FormulaEquationTagSplit(string.Empty, null);

        for (var index = trimmed.LastIndexOf("\\tag", StringComparison.Ordinal);
             index >= 0;
             index = trimmed.LastIndexOf("\\tag", index - 1, StringComparison.Ordinal))
        {
            var cursor = index + "\\tag".Length;
            if (cursor < trimmed.Length && trimmed[cursor] == '*') cursor++;
            if (cursor < trimmed.Length
                && trimmed[cursor] != '{'
                && !char.IsWhiteSpace(trimmed[cursor]))
                continue;
            while (cursor < trimmed.Length && char.IsWhiteSpace(trimmed[cursor])) cursor++;
            if (cursor >= trimmed.Length) continue;

            if (trimmed[cursor] == '{')
            {
                var close = MatchingClosingBrace(trimmed, cursor);
                if (close < 0 || trimmed.Substring(close + 1).Trim().Length != 0) continue;
                var tag = trimmed.Substring(cursor + 1, close - cursor - 1).Trim();
                if (tag.Length == 0) continue;
                return new FormulaEquationTagSplit(trimmed.Substring(0, index).TrimEnd(), tag);
            }

            var bareTag = trimmed.Substring(cursor).Trim();
            if (bareTag.Length == 0 || bareTag.IndexOfAny(new[] { '{', '}' }) >= 0) continue;
            return new FormulaEquationTagSplit(trimmed.Substring(0, index).TrimEnd(), bareTag);
        }

        return new FormulaEquationTagSplit(trimmed, null);
    }

    public static string Attach(string? source, string? equationTag)
    {
        var body = Extract(source).Latex;
        var tag = equationTag?.Trim();
        return string.IsNullOrWhiteSpace(tag) ? body : $"{body}\\tag{{{tag}}}";
    }

    private static int MatchingClosingBrace(string source, int openIndex)
    {
        var depth = 0;
        for (var index = openIndex; index < source.Length; index++)
        {
            if (source[index] == '\\')
            {
                index++;
                continue;
            }
            if (source[index] == '{') depth++;
            else if (source[index] == '}')
            {
                depth--;
                if (depth == 0) return index;
            }
        }
        return -1;
    }
}

File: apps/windows/src/office/shared/formulaEquationTag.ts
export interface FormulaEquationTagSplit {
  latex: string;
  equationTag: string | null;
}

function matchingClosingBrace(source: string, openIndex: number) {
  let depth = 0;
  for (let index = openIndex; index < source.length; index += 1) {
    const character = source[index];
    if (character === "\\") {
      index += 1;
      continue;
    }
    if (character === "{") depth += 1;
    else if (character === "}") {
      depth -= 1;
      if (depth === 0) return index;
    }
  }
  return -1;
}

/**
 * Extract a trailing equation tag from an editable formula body.
 *
 * Both canonical `\\tag{4.8.4}` and the malformed-but-common
 * `\\tag 4.8.4` form are accepted so older batch-import metadata can be
 * repaired without exposing the tag command to MathLive.
 */
export function splitFormulaEquationTag(source: string): FormulaEquationTagSplit {
  const trimmed = source.trim();
  if (!trimmed) return { latex: "", equationTag: null };

  for (let index = trimmed.lastIndexOf("\\tag"); index >= 0; index = trimmed.lastIndexOf("\\tag", index - 1)) {
    let cursor = index + "\\tag".length;
    if (trimmed[cursor] === "*") cursor += 1;
    const boundary = trimmed[cursor];
    if (boundary && boundary !== "{" && !/\s/.test(boundary)) continue;
    while (/\s/.test(trimmed[cursor] ?? "")) cursor += 1;
    if (cursor >= trimmed.length) continue;

    if (trimmed[cursor] === "{") {
      const close = matchingClosingBrace(trimmed, cursor);
      if (close < 0 || trimmed.slice(close + 1).trim()) continue;
      const equationTag = trimmed.slice(cursor + 1, close).trim();
      if (!equationTag) continue;
      return {
        latex: trimmed.slice(0, index).trimEnd(),
        equationTag,
      };
    }

    const equationTag = trimmed.slice(cursor).trim();
    if (!equationTag || /[{}]/.test(equationTag)) continue;
    return {
      latex: trimmed.slice(0, index).trimEnd(),
      equationTag,
    };
  }

  return { latex: trimmed, equationTag: null };
}

export function attachFormulaEquationTag(
  source: string,
  equationTag: string | null | undefined,
) {
  const body = splitFormulaEquationTag(source).latex;
  const tag = equationTag?.trim();
  return tag ? `${body}\\tag{${tag}}` : body;
}
